File: src/main/clj/design_patterns_with_clojure/chapter02/file_io_2.clj
(ns design-patterns-with-clojure.chapter02.file-io-2
  (:import
   (java.io
    FileInputStream
    FileOutputStream)
   (java.util
    Properties)))

(defprotocol FileIO
  (read-from-file [this filename])
  (write-to-file [this filename])
  (set-value [this key value])
  (get-value [this key]))

(extend-protocol FileIO
  Properties
  (read-from-file [this filename]
    (.load this (FileInputStream. filename)))
  (write-to-file [this filename]
    (.store this (FileOutputStream. filename)
            "written by FileProperties"))
  (set-value [this key value]
    (.setProperty this key value))
  (get-value [this key]
    (.getProperty this key)))

(defn -main []
  ;; protocol extension
  (doto (Properties.)
    (read-from-file "file.txt")
    (set-value "month" "4")
    (set-value "day" "21")
    (write-to-file "newfile1.txt")
    (set-value "year" "2004")
    (write-to-file "newfile2.txt")))

File: src/main/scala/design_patterns_with_scala/chapter02/FileIO2.scala
package design_patterns_with_scala.chapter02

import java.io.{FileInputStream, FileOutputStream}
import java.util.Properties

trait FileIO2[A]:
  extension (a: A)
    def readFromFile(filename: String): Unit
    def writeToFile(filename: String): Unit
    def setValue(key: String, value: String): Unit
    def getValue(key: String): String

given FileIO2[Properties] with
  extension (p: Properties)
    def readFromFile(filename: String): Unit =
      p.load(new FileInputStream(filename))

    def writeToFile(filename: String): Unit =
      p.store(new FileOutputStream(filename), "written by FileProperties")

    def setValue(key: String, value: String): Unit =
      p.setProperty(key, value)

    def getValue(key: String): String =
      p.getProperty(key)

@main
def fileIO2Example() =
  // instance of type class with extension methods
  val p = new Properties
  p.readFromFile("file.txt")
  p.setValue("month", "4")
  p.setValue("day", "21")
  p.writeToFile("newfile1.txt")
  p.setValue("year", "2004")
  p.writeToFile("newfile2.txt")
